bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryClientImpl.java

```diff
@@ -0,0 +1,492 @@
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.exceptions.ServiceFailureException;
+import org.apache.iceberg.exceptions.ServiceUnavailableException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/** A client of Google Bigquery Metastore functions over the BigQuery service. */
+public final class BigQueryClientImpl implements BigQueryClient {
+
+  public static final String NOT_AUTHORIZED_ERROR_MESSAGE =
+      "Not authorized to call the BigQuery API or access this resource";
+  private final Bigquery client;
+
+  /** Constructs a client of the Google BigQuery service. */
+  public BigQueryClientImpl() throws IOException, GeneralSecurityException {
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests
+    HttpCredentialsAdapter httpCredentialsAdapter =
+        new HttpCredentialsAdapter(
+            GoogleCredentials.getApplicationDefault().createScoped(BigqueryScopes.all()));
+    this.client =
+        new Bigquery.Builder(
+                GoogleNetHttpTransport.newTrustedTransport(),
+                GsonFactory.getDefaultInstance(),
+                httpRequest -> {
+                  httpCredentialsAdapter.initialize(httpRequest);
+                  httpRequest.setThrowExceptionOnExecuteError(
+                      false); // Less catching of exceptions, more analysis of the same HttpResponse
+                  // object, inspecting its status code
+                })
+            .setApplicationName("BigQuery Iceberg Catalog Plugin")
+            .build();
+  }
+
+  @VisibleForTesting
+  BigQueryClientImpl(Bigquery client) {
+    this.client = client;
+  }
+
+  @Override
+  public Dataset createDataset(Dataset dataset) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .insert(dataset.getDatasetReference().getProjectId(), dataset)
+              .executeUnparsed();
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  @Override
+  @SuppressWarnings("FormatStringAnnotation")
+  public Dataset getDataset(DatasetReference datasetReference) {
+    try {
+      HttpResponse response =
+          client
+              .datasets()
+              .get(datasetReference.getProjectId(), datasetReference.getDatasetId())
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(response.getStatusMessage());
```

Comment:
What is the status message? Is it the table name? What about the "full message" that is in the message content? Seems like handling the response should be the same across all places that throw exceptions.